### PR TITLE
Added support for serial MULTI protocol (serialMode == 5).

### PR DESCRIPTION
**Pull request by mikeller**

### CODE REVIEW

File: TX.h [hunk 1]
```diff
@@ -559,6 +566,55 @@ static inline void processSBUS(uint8_t c)
...
+        PPM[(set << 3) + 7] = calculateChannel(ppmWork.sbus.ch[set].ch7);
+      }
+      if ((ppmWork.sbus.status & 0x08) == 0) {
+#ifdef DEBUG_DUMP_PPM
+        ppmDump = 1;
+#endif
+        ppmAge = 0;
+      }
+
+      frameIndex = 0;
```
Comment (universam1) [inline TX.h]: Remove this line

Comment (mikeller) [inline TX.h]: Why?

Comment (universam1) [inline TX.h]: Sorry my error

**Comment by mikeller** [comment]
This has been confirmed working, and is good to go. Range now properly scales to 1000 to 2000. Configurator support in https://github.com/openLRSng/openLRSng-configurator/pull/30.

**Comment by mikeller** [comment]
Just had a test flight with ersky9xProvR219t_music from http://openrcforums.com/forum/viewtopic.php?f=7&t=4676. Serial RC command works fine, and telemetry works fine too with the changes in ersky9x to set the serial port used for telemetry to the same settings that are used for the serial RC output pin.

**Comment by universam1** [comment]
Thanks for testing @mikeller !
I will test with OpenTX asap. 

Can you monitor/ connect TX to PC for debug still since the serial is occupied now?

**Comment by mikeller** [comment]
Nope, unfortunately not.

**Comment by universam1** [comment]
Did ever someone try http://www.pjrc.com/teensy/td_libs_AltSoftSerial.html with native pins if that could potentially work? I think the RFM2x has some fifo buffer to give us some freedom

**Comment by mikeller** [comment]
@universam1: There's also this one:  http://www.hobbiesfly.com/transmitter-receiver/dtf-uhf-deluxetx-transmitter.html From reading the source it seems to have an additonal serial port (potentially as USB VCP), so this could be used for debugging.

**Comment by universam1** [comment]
yep TX_BOARD_TYPE 6 (Arduino Leonardo) has USB

**Comment by universam1** [comment]
I'm wild thinking, would it be possible and fast enough to use the original pin D3 with Soft/AltSerial also, instead of soldering to RX pin... Would give us access to config again.

**Comment by mikeller** [comment]
Interesting thought. Not having to do soldering would be a big advantage I think.

**Comment by universam1** [comment]
I've hacked it together and it does compile https://github.com/universam1/openLRSng/tree/Multi_with_SoftSerial
Unfortunately I cant test currently, I dont know if you would like to do a shot?

**Comment by DTFUHF** [comment]
So I've tried this and it looks good. The makefile changes update the required Arduino IDE to 1.6.11 which I haven't tested, but I have built this with Arduino IDE 1.6 (non-makefile).

Once the OLRS ID is included in the MULTI protocol, I think this will be ready to pull in.

**Comment by mikeller** [comment]
The OLRS id is available: https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/blob/master/Multiprotocol/Multiprotocol.h#L559

I will wait until I get a test version of the controller firmware that implements it, so I can test, and then I will switch the protocol .

**Comment by mikeller** [comment]
Ok, I changed the protocol Id to `0x1b`, tested with er9x (setting protocol to '26'), works fine, including telemetry.

**Comment by csurf** [comment]
I was talking to kha on IRC about supporting regular SBUS over softserial. This would be a good option for UHF/2.4 repeater setups where you have access to SBUS output from the 2.4 RX. By running SBUS input over softserial instead of UART0, the serial port on the TX is freed up for USB configuration, telemetry, or whatever else. Kha seemed to be OK with the idea, but he'd have to chime in and give his blessing.

I guess we'd need to add one more serialMode type'6'  to the list, "SBUS SS" (or some other way to indicate that it's normal SBUS over software serial).  Also, the existing code routines need to be adjusted to read normal SBUS from softserial in more/less same manner as MULTIPROTO is processed (wondering if existing MP code can be leveraged?).

**Comment by universam1** [comment]
@csurf You might take a look at my test branch, there you'll find a fine tuned SoftSerial implementation. SBUS should be no problem as any other supported protocol.

Though I would suggest instead of creating another protocol like SS SBUS rather to create a new config directive that defines the input way, so one of 3: "PPMin", "HardwareSerial" and "SoftSerial over PPMin". This way you have more freedom.

**Comment by csurf** [comment]
@universam1,
ok, I'll check your code out. I've just finished building my repeater, so I need to get the SBUS-softserial setup working now.

As far as the config directive goes, that sounds fine to me, as long as it's something that be configured from within configurator (i.e., doesn't require a rebuild). 

I like the idea of narrowing down & logically re-organizing the input options as you've suggested. Graphically, I guess there could be an option / drop-down box within configurator that allows you to select the appropriate serial protocol (SBUS, SUMD, MULTI, etc.) whenever one of the serial input methods is selected. The list would be hidden/disabled when PPMin is selected. This should also take care of selecting the right baud rate, so we can remove the serial protocols from the 'baud rate' settings (baud rate settings should ideally be disabled too if 'hardware serial' is selected as an input).

However, all of this assumes that the underlying serial processing must be able to transparently switch between hardware & software serial independent of what serial protocol is selected. Is this currently how it works with your code? (sorry, haven't had a chance to look at it yet)

BTW, should I push this stuff off into a new issue, or continue to work with it here?

**Comment by mikeller** [comment]
@csurf Totally agree with you that a split of the current 'serial speed' setting into 'protocol' / 'speed' (and potentially 'port' would be an improvement.

However, re the soft serial code, it has to be said that until now it has not been shown that this is even feasible, and if it is, if it is possible to throttle the input rate in such a way on the module running openLRSng is not running the danger of being blown out of the water if the hardware sending the serial control signal isn't throttling the data being sent to leave enough silence on the line to give the openLRSng MCU time to execute the protocol loop.

Also, can you please explain why you would want to run a repeater by using an emulated serial port, instead of using the much more suitable hardware serial port for the important signal (RC data), and if needed emulating a serial port for less important data like telemetry?

**Comment by csurf** [comment]
@universam1,
ok, I downloaded your multiprotocol test branch, and could not get your software serial implementation to work. I'm using Arduino IDE 1.0.6, board type 4 (DTF UHF Deluxe V2), non-inverted SBUS on PPM pin (I'm using an external hardware inverter to correct the SBUS signal from an X8R). 

I tried debugging in a few different ways, but I'm not having any luck. I tried setting the softserial mode to 8E2, but it didn't help. Don' know what else to try. Any ideas?

@mikeller, reason for softser was mainly for initial testing, in order to inspect the raw, incoming SBUS data from the RX. There are a couple of other quirky issues that I've run into due to using hardware serial input, such as slightly longer boot times and trouble with going into bind mode. It also probably interferes with USB connection, but I don't know enough about the technical details to say for sure. Obviously, I could just disconnect the S.BUS input when I need to bind or go into USB, but I figured it might be worthwhile to explore the softser option. So far, it's not looking good though...

**Comment by mikeller** [comment]
@csurf: From reading the source, I think the DTFUHF deluxe has an MCU that has two serial ports, so you should be able to debug on one with a serial connection on the other.

**Comment by csurf** [comment]
@mikeller, 
I have the deluxe V2, which is similar to the JR module (ATMega328P & CP2102). So, only one hardware serial port unfortunately...

**Comment by universam1** [comment]
@csurf It's hard to tell why you have problems here, since I tested only Multi which is though a SBUS derivative protocol. It's tuned to 8E2 that fits to SBUS.... Have you tried some debugging of RAW values?

**Comment by mikeller** [comment]
Superseded by #122.

**Pull request closed without merging**